wordpress-mobile/WordPress-Android
# Gutenberg - async media upload Post background processing

Fixes part of https://github.com/wordpress-mobile/gutenberg-mobile/issues/475

Follow up to #9129 (media upload progress reattach). That PR handles the visual updates, while this other PR handles the updates to a Gutenberg Post when an image upload finishes but the user is _not currently using the Gutenberg editor_.

This PR introduces a `MediaUploadReadyProcessor` implementation for Gutenberg `Image` blocks based on search & replace, and also makes Post<-> media associations aware of Gutenberg posts in `purgeMediaToPostAssociationsIfNotInPostAnymore`.

For the first (`MediaUploadReadyProcessor`), I've been considering some possible paths and have made up my mind for the easiest to implement for now, but may as well not be the most reliable as it makes several assumptions regarding the structure of an Image block as it currently is.

However, the PR works as per the current state of affairs, and I believe we shouldn't expect there to be breaking changes to the block structure in the future given it's one of the core blocks.


#### For further discussion
Other possibilities:
- Use React Native's headless JS https://facebook.github.io/react-native/docs/headless-js-android and get an instance of Gutenberg parser there somehow (this is a path that needs more investigation), special attention to this:

> By default, your app will crash if you try to run a task while the app is in the foreground. This is to prevent developers from shooting themselves in the foot by doing a lot of work in a task and slowing the UI. You can pass a fourth boolean argument to control this behaviour.


(this is indeed to be used only when the GB Editor is not on the foreground, but we should also consider situations in which we have already launched the background task and the user is also trying to bring GB editor up in the middle.)

- use an HTML parser at least to make sense of the HTML tags to make url replacement a little bit more intelligent (or, HTML aware)

- use some kind of regex or string replacement like this one here.
 

#### To test:
Pretty much we can follow the test cases described in https://github.com/wordpress-mobile/WordPress-Android/pull/5780, writing here the most simple case (situation described in CASE B there):

1. start a new draft
2. insert an image block and choose an from device
3. tap the back arrow to exit the editor, observe the image continues to upload
4. also tap back again to exit the Posts list, back to the main app screen (as the posts list screen gets updated more frequently and can change the effect of this test). Even send the app to the background if you want.
5. once the image is done uploading a notification will appear.
6. go to the web and open the Post
7. confirm the image block has no validation errors, and the Post opens and displays the image correctly.

Update release notes:
